docs/how-to/hip_cpp_language_extensions.rst

```diff
@@ -432,6 +432,149 @@ compile-time constant on the host. It has to be queried using
+in the following example.
+
+.. code-block:: cpp
+  :emphasize-lines: 17, 18, 61-64, 80-88
+
+  #include <hip/hip_runtime.h>
+  #include <iostream>
+  #include <vector>
+
+  #define HIP_CHECK(expression)                \
+  {                                            \
+      const hipError_t status = expression;    \
+      if(status != hipSuccess){                \
+              std::cerr << "HIP error "        \
+                  << status << ": "            \
+                  << hipGetErrorString(status) \
+                  << " at " << __FILE__ << ":" \
+                  << __LINE__ << std::endl;    \
+      }                                        \
+  }
+
+  template<uint32_t WarpSize>
+  __global__ void block_reduce(int* input, int* output, size_t size){
+    extern __shared__ int shared[];
+
+    // Overindex-safe read of input
+    auto read_global_safe = [&](const uint32_t i)
+    {
+      return i < size ? input[i] : 0;
+    };
+
+    const uint32_t tid = threadIdx.x,
+                   bid = blockIdx.x,
+                   gid = bid * blockDim.x + tid;
+
+    // Read input buffer to shared
+    shared[tid] = read_global_safe(gid);
+    __syncthreads();
+
+    // Shared reduction
+    for (uint32_t i = blockDim.x / 2; i >= WarpSize; i /= 2)
+    {
+      if (tid < i)
+        shared[tid] = shared[tid] + shared[tid + i];
+      __syncthreads();
+    }
+  
+    // Use local variable in warp reduction  
+    int result =  shared[tid];
+    __syncthreads();
+
+    #pragma unroll
+    for (uint32_t i = WarpSize/2; i != 0; i /= 2) {
+      if (tid < i) {
+        result = result + __shfl_up(result, i);
+	  }
+    }
+	
+    // Write result to output buffer
+    if (tid == 0)
+      output[bid] = result;
+  };
+  
+  int main() {
+
+    int deviceId = 0;
+    int warpSizeHost;
+    HIP_CHECK(hipDeviceGetAttribute(&warpSizeHost, hipDeviceAttributeWarpSize, deviceId));
+    std::cout << "Warp size: " << warpSizeHost << std::endl;
+
+    constexpr int numOfBlocks = 16;
+    constexpr int threadsPerBlock = 1024;
+    constexpr size_t arraySize = numOfBlocks * threadsPerBlock;
+
+    int *d_data, *d_results;
+    int initValue = 1;
+    std::vector<int> vectorInput(arraySize, initValue);
+    std::vector<int> vectorOutput(numOfBlocks);
+    // Allocate device memory
+    HIP_CHECK(hipMalloc(&d_data, arraySize * sizeof(*d_data)));
+    HIP_CHECK(hipMalloc(&d_results, numOfBlocks * sizeof(*d_results)));
+    // Host to Device copy of the input array
+    HIP_CHECK(hipMemcpy(d_data, vectorInput.data(), arraySize * sizeof(*d_data), hipMemcpyHostToDevice));
+    
+    // Select the kernel
+    if(warpSizeHost == 32) {
+      block_reduce<32><<<dim3(numOfBlocks), dim3(threadsPerBlock), threadsPerBlock * sizeof(*d_data)>>>(d_data, d_results, arraySize);
+    } else if(warpSizeHost == 64) {
+      block_reduce<64><<<dim3(numOfBlocks), dim3(threadsPerBlock), threadsPerBlock * sizeof(*d_data)>>>(d_data, d_results, arraySize);
+    } else {
```

Comment:
As warpSizeHost is determined by hipDeviceAttributeWarpSize is it necessary to cover the "Unsupported warp size" here?